### ISSUE TITLE
Implement helper function to recursive compose objects.

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp#3016]
The JSON API is limited to composing 32 objects at a time:

https://cloud.google.com/storage/docs/json_api/v1/objects/compose

However there is no limit to the number of components in an object:

https://cloud.google.com/storage/docs/composite-objects#_Count

The library could have a helper function to efficiently compose "N" objects where N is much larger than 32.  The library should divide the source objects is groups of 32 or less objects, then compose those objects (in parallel?) once those succeed it can recursively compose the resulting objects again until there is only one object left to compose.

**Comment by acastruc** [comment]
Not trying to press my luck, but it would be awesome if this function had an option to delete the parts?

**Comment by coryan** [comment]
I think this will have to remove the temporary objects it creates, so adding an option to remove the initial files does not sound that out of scope.

I do worry about the return type for this.  The compose operations may fail (even after retries) and we should stop merging, delete *only* the temporary objects in that case.  But those deletes may also fail, and we need to indicate which ones so the application developers can delete those themselves (and/or they should be created with a lifecycle policy to be automatically deleted by the service).

Only once the composition has succeeded we can proceed to delete the source objects. But that *also* fail, so we should report any failures in that step.  The return value is becoming something like:

```C++
namespace google::cloud::storage {

struct CompositionResult {
  StatusOr<ObjectMetadata> composed_object;
  // Object names and reason for delete failures
  std::vector<std::pair<std::string, Status>> delete_failures;
};

// Standalone function to compose multiple objects.
template <....>
CompositionResult ComposeMany(
    std::string bucket_name,
    std::vector<ComposeSourceObject> source_objects,
    std::string destination_object_name,
    Options&&... options);
}  // namespace 
```

Because this operation may take "a while" to run, we should consider returning this wrapped in a `google::cloud::future<>`.

**Comment by coryan** [comment]
@dopiera we should discuss this in the next meeting.

**Comment by dopiera** [comment]
I think we should take a different approach.

The client process might simply die in the middle of the operation. With this proposal we won't have an easy way to communicate to the user which temporary objects can be deleted. Similarly, if the client dies after executing the final composition but before removing the source files, a manual user intervention is still necessary.

It propose that we add a mandatory `prefix` parameter to `ComposeMany()`. By specifying it, the user promises that no object with this prefix will be created by any other operations. We then implement a helper `DeleteByPrefix()` to do a manual cleanup in case `ComposaMany()` fails to delete some files.

Assuming this functionality is mainly going to be used for parallel uploads, one could even put the parts of the upload and the temporary, partially composed files under a shared prefix so that both source files and the temporary files are easily trackable and deletable via a single `DeleteByPrefix()`.

Thoughts?

**Comment by coryan** [comment]
> I think we should take a different approach.

I think I like your approach, but let me write something down to make sure we agree.

I think you are suggesting that `ComposeMany` looks like this:

```C++
// Standalone function to compose multiple objects.
template <....>
CompositionResult ComposeMany(
    Client client,
    std::string bucket_name,
    std::vector<ComposeSourceObject> source_objects,
    std::string destination_object_name,
    std::string prefix_for_temporaries,
    Options&&... options);
```

and that we *require* that `prefix_for_temporaries` is use exclusively by that particular invocation of  `ComposeMany()`, so we can safely delete all the temporaries with that prefix.

Did I capture your intent so far?


And then you are also proposing a helper function:

```C++
template <....>
void DeleteByPrefix(
    Client client,
    std::string bucket_name,
    std::string prefix,
    Options&&... options);
```

> It propose that we add a mandatory `prefix` parameter to `ComposeMany()`. By specifying it, the user promises that no object with this prefix will be created by any other operations.

Maybe we could add a helper function to create those?  An equivalent to `mkstemp` that creates a random name and puts a marker object (with `IfGenerationMatch(0)`) to ensure no other thread / process will use it?

> 
> Assuming this functionality is mainly going to be used for parallel uploads, one could even put the parts of the upload and the temporary, partially composed files under a shared prefix so that both source files and the temporary files are easily trackable and deletable via a single `DeleteByPrefix()`.
> 
> Thoughts?

Assuming I did not misunderstand anything: I like it, go for it.

When you create those temporary objects, remember that object lifecycle management can make them automatically temporary for you....

**Comment by dopiera** [comment]
> ```c++
> // Standalone function to compose multiple objects.
> template <....>
> CompositionResult ComposeMany(
>     Client client,
>     std::string bucket_name,
>     std::vector<ComposeSourceObject> source_objects,
>     std::string destination_object_name,
>     std::string prefix_for_temporaries,
>     Options&&... options);
> ```
> 
> and that we _require_ that `prefix_for_temporaries` is use exclusively by that particular invocation of `ComposeMany()`, so we can safely delete all the temporaries with that prefix.
> 
> Did I capture your intent so far?

Yes, that's the idea.

I didn't write it, but I also thought that we could simply return `StatusOr<ObjectMetadata>` to make the API slightly more friendly. The operation would succeed even if deleting any of the temporary files failed. The reasoning behind it is that the user would have to be prepared for some leftovers anyway  and they'd have an expiration of e.g. hour. I'm not super sure if this is a good idea, though.

> And then you are also proposing a helper function:
> 
> ```c++
> template <....>
> void DeleteByPrefix(
>     Client client,
>     std::string bucket_name,
>     std::string prefix,
>     Options&&... options);
> ```
Yes.

> > It propose that we add a mandatory `prefix` parameter to `ComposeMany()`. By specifying it, the user promises that no object with this prefix will be created by any other operations.
 
> Maybe we could add a helper function to create those? An equivalent to `mkstemp` that creates a random name and puts a marker object (with `IfGenerationMatch(0)`) to ensure no other thread / process will use it?

I didn't think about it, but yes, absolutely.
 
> Assuming I did not misunderstand anything: I like it, go for it.

Awesome, will do.

> When you create those temporary objects, remember that object lifecycle management can make them automatically temporary for you....

Yup.

**Comment by dopiera** [comment]
FWIW, the implementation is on my branch in these commits:
* implement std::tuple filtering based on types:
https://github.com/dopiera/google-cloud-cpp/commit/97a46735a59dc49c80c73c68f8318944a18fd26c
* implement `DeleteByPrefix`: https://github.com/dopiera/google-cloud-cpp/commit/bd75ad53f05cdbfd3f32557357b4e9bb3cc08b38
* implement actual `ComposeMany` https://github.com/dopiera/google-cloud-cpp/commit/60fa4402bb317c56a6c8cea6e558119c7024bcb7

I need to cut a release in `google-cloud-cpp-common` before merging, because it contains the crucial reimplementation of `std::apply`: https://github.com/googleapis/google-cloud-cpp-common/commit/6fe4477232a487e1d2371633a87a9db3a5af6f87
